### PR TITLE
[Don't merge][test] mir niche checks

### PR DESCRIPTION
**Pull request by dianqk**
This PR is used to test #104862.
Upstream patch: https://github.com/llvm/llvm-project/pull/68190.
Discussion: https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Implementing.20niche.20checks.

r? saethlin

### CODE REVIEW

File: compiler/rustc_mir_transform/src/check_niches.rs [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+    fn insert_niche_check(
+        &mut self,
+        block_data: &mut BasicBlockData<'tcx>,
+        new_block: BasicBlock,
+        operand: Operand<'tcx>,
+        niche: NicheKind,
+        source_info: SourceInfo,
+    ) {
+        let type_name =
+            with_no_trimmed_paths!(format!("{:?}", operand.ty(self.local_decls, self.tcx)));
```
Comment (bjorn3) [inline compiler/rustc_mir_transform/src/check_niches.rs:119]: Is there a reason you are not letting the codegen backend do this? That should reduce the crate metadata size by avoiding a string to be stored.

Comment (DianQK) [reply compiler/rustc_mir_transform/src/check_niches.rs:119]: Maybe you should comment at #104862? :p

We found a mis-compilation in this PR. This PR is used to validate the fix.

**Comment by saethlin** [comment]
@DianQK If you want to get CI to pass, you probably need to add `-Zmir-enable-passes=-CheckNiches` or `-Cdebug-assertions=no` to the tests that are failing due to perturbation by the check.

**Comment by DianQK** [comment]
I just remembered that `x86_64-gnu-llvm-15` and `x86_64-gnu-tools` do not use the built-in LLVM.

**Comment by DianQK** [comment]
Verify completion.

**Comment by DianQK** [comment]
Check again due to the reimplementation of the upstream patch.

**Pull request closed without merging**